File: camerax_platform_interface/lib/src/core/camera_info.dart
import 'package:camerax_platform_interface/src/common.dart';
import 'package:camerax_platform_interface/src/core.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

abstract interface class CameraInfo {
  static Future<bool> mustPlayShutterSound() =>
      CameraInfoChannel.instance.mustPlayShutterSound();

  Future<CameraSelector> getCameraSelector();
  Future<LiveData<CameraState>> getCameraState();
  Future<ExposureState> getExposureState();
  Future<double> getIntrinsicZoomRatio();
  Future<CameraSelector$LensFacing> getLensFacing();
  Future<LiveData<LowLightBoostState>> getLowLightBoostState();
  Future<int> getMaxTorchStrengthLevel();
  Future<Set<CameraInfo>> getPhysicalCameraInfos();
  Future<int> getSensorRotationDegrees([Surface$Rotation? relativeRotation]);
  Future<Set<Range<int>>> getSupportedFrameRateRanges();
  Future<LiveData<TorchState>> getTorchState();
  Future<LiveData<int>> getTorchStrengthLevel();
  Future<LiveData<ZoomState>> getZoomState();
  Future<bool> hasFlashUnit();
  Future<bool> isFocusMeteringSupported(FocusMeteringAction action);
  Future<bool> isLogicalMultiCameraSupported();
  Future<bool> isLowLightBoostSupported();
  Future<bool> isTorchStrengthSupported();
  Future<bool> isZslSupported();
  Future<Set<DynamicRange>> querySupportedDynamicRanges(
    Set<DynamicRange> candidateDynamicRanges,
  );
}

abstract base class CameraInfoChannel extends PlatformInterface {
  CameraInfoChannel() : super(token: _token);

  static final _token = Object();

  static CameraInfoChannel? _instance;

  static CameraInfoChannel get instance =>
      ArgumentError.checkNotNull(_instance, 'instance');

  static set instance(CameraInfoChannel instance) {
    PlatformInterface.verify(instance, _token);
    _instance = instance;
  }

  Future<bool> mustPlayShutterSound();
}

File: camerax_ios/ios/camerax_ios/Sources/camerax_ios/core/CameraInfo.swift
//
//  CameraInfo.swift
//  camerax_ios
//
//  Created by 闫守旺 on 2025/5/22.
//

import Foundation
import AVFoundation

public class CameraInfo: NSObject {
    private var cameraSelector: CameraSelector
    private let torchControl: TorchControl
    private let zoomControl: ZoomControl
    
    init(cameraSelector: CameraSelector, torchControl: TorchControl, zoomControl: ZoomControl) {
        self.cameraSelector = cameraSelector
        self.torchControl = torchControl
        self.zoomControl = zoomControl
    }
    
    public func getCameraSelector() -> CameraSelector {
        return self.cameraSelector
    }
    
    func setCameraSelector(_ cameraSelector: CameraSelector) {
        self.cameraSelector = cameraSelector
    }
    
    public func getTorchState() -> LiveData<TorchState> {
        return self.torchControl.getTorchState()
    }
    
    public func getZoomState() -> LiveData<ZoomState> {
        return self.zoomControl.getZoomState()
    }
}

public extension CameraInfo {
    func getLensFacing() -> CameraSelector.LensFacing {
        let cameraSelector = getCameraSelector()
        return cameraSelector.lensFacing
    }
}
